libnetwork/drivers/bridge/setup_ip_tables_linux.go

```diff
@@ -191,88 +192,153 @@ func setupIPChains(config configuration, version iptables.IPVersion) (retErr err
+	Hairpin    bool
+	Enable4    bool
+	Enable6    bool
+}
+
+type iptablesNetwork struct {
+	networkConfig
+	cleanFuncs iptablesCleanFuncs
+}
+
+func newIptablesNetwork(nc networkConfig) (_ *iptablesNetwork, retErr error) {
+	n := &iptablesNetwork{
+		networkConfig: nc,
+	}
+	defer func() {
+		if retErr != nil {
+			n.delNetworkLevelRules()
+		}
+	}()
+
+	if err := n.reapplyNetworkLevelRules(); err != nil {
+		return nil, err
 	}
-	return n.setupIPTables(iptables.IPv4, maskedAddrv4, config, i)
+	return n, nil
 }
 
-func (n *bridgeNetwork) setupIP6Tables(config *networkConfiguration, i *bridgeInterface) error {
-	maskedAddrv6 := &net.IPNet{
-		IP:   i.bridgeIPv6.IP.Mask(i.bridgeIPv6.Mask),
-		Mask: i.bridgeIPv6.Mask,
+func (n *iptablesNetwork) reapplyNetworkLevelRules() error {
+	if n.Enable4 {
+		if err := n.configure(iptables.IPv4, n.Config4); err != nil {
+			return err
+		}
 	}
-	return n.setupIPTables(iptables.IPv6, maskedAddrv6, config, i)
+	if n.Enable6 {
+		if err := n.configure(iptables.IPv6, n.Config6); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (n *bridgeNetwork) setupIPTables(ipVersion iptables.IPVersion, maskedAddr *net.IPNet, config *networkConfiguration, i *bridgeInterface) error {
-	var err error
+func (n *iptablesNetwork) delNetworkLevelRules() error {
+	var errs []error
+	for _, cleanFunc := range n.cleanFuncs {
+		if err := cleanFunc(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	n.cleanFuncs = nil
+	return errors.Join(errs...)
+}
 
-	d := n.driver
-	d.Lock()
-	driverConfig := d.config
-	d.Unlock()
+func (n *iptablesNetwork) configure(ipv iptables.IPVersion, conf networkConfigFam) error {
+	if !conf.Prefix.IsValid() {
+		// Delete INC rules, in case they were created by a 28.0.0 daemon that didn't check
+		// whether the network had iptables/ip6tables enabled.
+		// This preserves https://github.com/moby/moby/commit/8cc4d1d4a2b6408232041f9ba4dff966eba80cc0
+		return setINC(ipv, n.IfName, conf.Routed, false)
+	}
+	if err := n.setupIPTables(ipv, conf); err != nil {
+		return err
+	}
+	return nil
+}
 
-	// Pickup this configuration option from driver
-	hairpinMode := !driverConfig.EnableUserlandProxy
+func (n *iptablesNetwork) registerCleanFunc(clean iptableCleanFunc) {
+	n.cleanFuncs = append(n.cleanFuncs, clean)
+}
 
-	if config.Internal {
-		if err = setupInternalNetworkRules(config.BridgeName, maskedAddr, config.EnableICC, true); err != nil {
+func (n *iptablesNetwork) setupIPTables(ipVersion iptables.IPVersion, config networkConfigFam) error {
+	if n.Internal {
+		if err := setupInternalNetworkRules(n.IfName, config.Prefix, n.ICC, true); err != nil {
 			return fmt.Errorf("Failed to Setup IP tables: %w", err)
 		}
-		n.registerIptCleanFunc(func() error {
-			return setupInternalNetworkRules(config.BridgeName, maskedAddr, config.EnableICC, false)
+		n.registerCleanFunc(func() error {
+			return setupInternalNetworkRules(n.IfName, config.Prefix, n.ICC, false)
 		})
 	} else {
```

Comment:
For a follow-up, we could early-return here and de-intend the non-internal branch.